packages/ckeditor5-ui/src/menubar/utils.ts

```diff
@@ -163,6 +163,32 @@ export const MenuBarBehaviors = {
 			callback: () => menuBarView.close(),
 			contextElements: () => menuBarView.children.map( child => child.element! )
 		} );
+	},
+
+	/**
+	 * Tracks the keyboard focus interaction on the menu bar view. It is used to determine if the nested items
+	 * of the menu bar should render focus rings after first interaction with the keyboard.
+	 */
+	trackKeyboardFocusInteraction( menuBarView: MenuBarView ): void {
+		let isKeyPressed: boolean = false;
+
+		menuBarView.on<ObservableChangeEvent<boolean>>( 'change:isOpen', () => {
+			menuBarView.hadKeyboardFocusInteraction = false;
+		} );
+
+		menuBarView.listenTo( menuBarView.element!, 'keydown', () => {
+			isKeyPressed = true;
+		}, { useCapture: true } );
+
+		menuBarView.listenTo( menuBarView.element!, 'keyup', () => {
+			isKeyPressed = false;
+		}, { useCapture: true } );
+
+		menuBarView.listenTo( menuBarView.element!, 'focus', () => {
```

Comment:
I just accidentally lost a long comment, so I am super happy to write it again :|.

/cc @oleq

So, the idea here is that: if focus has changed and keyboard key was pressed, then it means that this keyboard interaction caused the focus change and we need to start displaying outline.

This is an interesting, simple and smart idea. But I am not sure if I like it in the end.

First, it does not work for enter key (see my other comment).

Second, it grabs false positives. E.g. if I have shift key pressed and start hovering over menus, the outline gets displayed.

If we don't find a better and still reasonably simple solution then I am fine with that. But maybe there's some room for improvement. @oleq if you think that current solution is good enough we can leave it as is.

So, we can go one of two directions:

1.  Try to improve what we have. Try to solve enter key bug and maybe introduce list of keys that trigger this behavior (will fix false positives but may introduce false negatives if we introduce a behavior for a new key and forget to update the list).
2.  Try to implement `hadKeyboardFocusInteraction` handling precisely when it needs to be handled. This means going through all behaviors functions, possibly propagating something "up" to the menu bar component. I also think there's no behavior function for "open on enter" behavior and that's wired to the `execute` event, which again may be difficult to handle. I am afraid this will take us some time and ROI may be too small.

Again, @oleq please let me know what you think.

Reply:
@scofalik Originally I implemented it using `KeyboardFocusTracker` that was observing every keyboard interaction on the document and then there were similar comparison for `isKeyPressed`. It worked fine, and it fixed the issue you mentioned with `enter` and `alt+f9`, and that one with focus jumping from another element to menu using keyboard, although after discussion it was not a great idea to add another global utility, and it'll be better to use it locally so I removed it. May it be better to keep this utility as global handler?

Reply:
I remember the original, implementation, hence I wrote:

> If we don't find a better and still reasonably simple solution then I am fine with that.

:slightly\_smiling\_face: 

In case of this particular topic (showing focus outline), keyboard interaction is a keyboard interaction that would lead to changing the menu state, including both focus and opening menus (and in that one case with enter, enter key does not change focus but I think "it counts").

But now as I think about it, the point where we should start tracking focus is actually Alt+F9 press. Because that's the first keyboard interaction that changes focus/menu state.

Isn't the solution as simple as just listening one element "higher", instead of on menu views, then on the menu bar?

Reply:
> But now as I think about it, the point where we should start tracking focus is actually Alt+F9 press. 

@scofalik Yep, I discovered that in later commits and it looks like it works. Regarding `enter` issue - it was issue related to incorrect reset of `isKeyPressed` flag because some of the menus tends to intercept `keyup` and it was never freed.